pkg/webhooks/validation/server/server.go

```diff
@@ -0,0 +1,500 @@
+
+	// Enable galley validation mode
+	Enabled bool
+
+	// Mux of an existing HTTP server, Galley will not manage its own
+	Mux *http.ServeMux
+}
+
+// String produces a stringified version of the arguments for debugging.
+func (p *Options) String() string {
+	buf := &bytes.Buffer{}
+
+	fmt.Fprintf(buf, "DomainSuffix: %s\n", p.DomainSuffix)
+	fmt.Fprintf(buf, "Port: %d\n", p.Port)
+	fmt.Fprintf(buf, "CertFile: %s\n", p.CertFile)
+	fmt.Fprintf(buf, "KeyFile: %s\n", p.KeyFile)
+
+	return buf.String()
+}
+
+// DefaultArgs allocates an Options struct initialized with Webhook's default configuration.
+func DefaultArgs() Options {
+	return Options{
+		Port:     9443,
+		CertFile: constants.DefaultCertChain,
+		KeyFile:  constants.DefaultKey,
+		Enabled:  true,
+	}
+}
+
+// Webhook implements the validating admission webhook for validating Istio configuration.
+type Webhook struct {
+	keyCertWatcher *fsnotify.Watcher
+
+	mu   sync.RWMutex
+	cert *tls.Certificate
+
+	// pilot
+	descriptor   schema.Set
+	domainSuffix string
+
+	// mixer
+	validator store.BackendValidator
+
+	server   *http.Server
+	keyFile  string
+	certFile string
+}
+
+// Reload the server's cert/key for TLS from file and save it for later use by the https server.
+func (wh *Webhook) reloadKeyCert() {
+	pair, err := reloadKeyCert(wh.certFile, wh.keyFile)
+	if err != nil {
+		return
+	}
+
+	wh.mu.Lock()
+	wh.cert = pair
+	wh.mu.Unlock()
+}
+
+// Reload the server's cert/key for TLS from file.
+func reloadKeyCert(certFile, keyFile string) (*tls.Certificate, error) {
+	pair, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		reportValidationCertKeyUpdateError(err)
+		scope.Warnf("Cert/Key reload error: %v", err)
+		return nil, err
+	}
+
+	reportValidationCertKeyUpdate()
+	scope.Info("Cert and Key reloaded")
+
+	var row int
+	for _, cert := range pair.Certificate {
+		if x509Cert, err := x509.ParseCertificates(cert); err != nil {
+			scope.Infof("x509 cert [%v] - ParseCertificates() error: %v\n", row, err)
+			row++
+		} else {
+			for _, c := range x509Cert {
+				scope.Infof("x509 cert [%v] - Issuer: %q, Subject: %q, SN: %x, NotBefore: %q, NotAfter: %q\n",
+					row, c.Issuer, c.Subject, c.SerialNumber,
+					c.NotBefore.Format(time.RFC3339), c.NotAfter.Format(time.RFC3339))
+				row++
+			}
+		}
+	}
+	return &pair, nil
+}
+
+// New creates a new instance of the admission webhook server.
+func New(p Options) (*Webhook, error) {
+	if p.Mux != nil {
+		wh := &Webhook{
+			descriptor: p.PilotDescriptor,
+			validator:  p.MixerValidator,
+		}
+
+		p.Mux.HandleFunc("/admitpilot", wh.serveAdmitPilot)
+		p.Mux.HandleFunc("/admitmixer", wh.serveAdmitMixer)
+		return wh, nil
+	}
+	pair, err := reloadKeyCert(p.CertFile, p.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configuration must be updated whenever the caBundle changes. Watch the parent directory of
+	// the target files so we can catch symlink updates of k8s secrets.
+	keyCertWatcher, err := fsnotify.NewWatcher()
```

Comment:
Seems like we use a different filewatcher library here and in the controller? Is that expected?

Reply:
Ah, this was left over from the original un-forked server.